### ISSUE TITLE
boost-1.75.0 port on kinvolk portage-stable

### ISSUE THREAD

**Issue by bonkko** [flatcar/Flatcar#362]
**Proposal**
Add _boost-1.75.0_ to stable portage tree.
**Impact**
Use of _Python 3_6_. 
**Ideal future situation**
The package should work on the same arch of _Gentoo_. 
**Additional information**
see https://github.com/kinvolk/portage-stable/pull/153 PR
see https://github.com/kinvolk/Flatcar/issues/362#issuecomment-801875240 as reference

**Comment by bonkko** [comment]
### Documentation of the proposal (first draft)
**Current situation**
I made a successful build of boost-1.75 with the boost-build taken from _https://github.com/boostorg/build_ **version 4.4.1**. I put the _compilation with -j1_ and I _disabled python_ because with _python flag enabled_ the  compilation fail. With _-.j1_ and _-python_ the compilation is slow but It finished without any fatal error.

```
>>> Installing (1 of 1) dev-libs/boost-1.75.0::coreos
 * Boost.Regex is *extremely* ABI sensitive. If you get errors such as
 * 
/var/tmp/portage/dev-libs/boost-1.75.0/temp/environment: line 3035: ver_cut: command not found
/var/tmp/portage/dev-libs/boost-1.75.0/temp/environment: line 3035: ver_cut: command not found
 *   undefined reference to `boost::re_detail_000::cpp_regex_traits_implementation
 *     <char>::transform_primary[abi:cxx11](char const*, char const*) const'
 * 
 * Then you need to recompile Boost and all its reverse dependencies
 * using the same toolchain. In general, *every* change of the C++ toolchain
 * requires a complete rebuild of the boost-dependent ecosystem.
 * 
 * See for instance https://bugs.gentoo.org/638138

 * Messages for package dev-libs/boost-1.75.0:
 * Log file: /var/log/portage/dev-libs:boost-1.75.0:20210309-150849.log

 * Boost.Regex is *extremely* ABI sensitive. If you get errors such as
 * 
 *   undefined reference to `boost::re_detail_000::cpp_regex_traits_implementation
 *     <char>::transform_primary[abi:cxx11](char const*, char const*) const'
 * 
 * Then you need to recompile Boost and all its reverse dependencies
 * using the same toolchain. In general, *every* change of the C++ toolchain
 * requires a complete rebuild of the boost-dependent ecosystem.
 * 
 * See for instance https://bugs.gentoo.org/638138
>>> Auto-cleaning packages...

>>> No outdated packages were found on your system.

 *GNU info directory index is up-to-date.
```



**Differences with Gentoo**

 **dev-util/boost-build-1.75.0**  _boost-build version 4.4.1 cloned with git_  and _EAPI6_ and _python3_6_
 
 **dev-libs/boost-1.75.0**

```
EAPI=6

PYTHON_COMPAT=(python3_6)
IUSE="bzip2 context debug doc icu lzma -nls mpi numpy -python static-libs +threads tools zlib zstd"
MY_PV="1_75_0"
MAJOR_V="1_75_0"
src_configure() {
        OPTIONS=(
                "-j1"
                -q
                -d+2
                pch=off
                $(usex icu "-sICU_PATH=${ESYSROOT}/usr" '--disable-icu boost.locale.icu=off')
                $(usex mpi '' '--without-mpi')
                $(usex nls '' '--without-locale')
                $(usex context '' '--without-context --without-coroutine --without-fiber')
                $(usex threads '' '--without-thread')
                --without-stacktrace
                --prefix="${ED}/usr"
                --layout=system
                # CMake has issues working with multiple python impls,
                # disable cmake config generation for the time being
                # https://github.com/boostorg/python/issues/262#issuecomment-483069294
                # building with threading=single is currently not possible
                # https://svn.boost.org/trac/boost/ticket/7105
                threading=multi
                link=$(usex static-libs shared,static shared)
                # this seems to be the only way to disable compression algorithms
                # https://www.boost.org/doc/libs/1_70_0/libs/iostreams/doc/installation.html#boost-build
                -sNO_BZIP2=$(usex bzip2 0 1)
                -sNO_LZMA=$(usex lzma 0 1)
                -sNO_ZLIB=$(usex zlib 0 1)
                -sNO_ZSTD=$(usex zstd 0 1)
        )
        append-cxxflags -std=c++11
}
```

**Additional information**
I do not make test for _aarch64_ and _ppc32_, and I do not make unit test. I would be happy to work to packages stabilization and test other arch with different compile flags. In my honest opinion flatcar can benefit from multiple arch support.

**How To**
This section is not definitive but it may help to make a _first test_. 

Clone my repo  
```
git clone https://github.com/bonkko/portage-stable.git
git checkout boost-1.75.0-port
```
Copy the ebuild in the right dir and make the Manifest
```
ebuild build-boost-1.75.0.ebuild digest
ebuild boost-1.75.0.ebuild digest
```
Run emerge in the ebuild directory
```
emerge build-boost-1.75.0.ebuild 
emerge boost-1.75.0.ebuild
```
To unmask the packages and use emerge as normal you should add comment in the following line
```
nano /mnt/host/source/src/third_party/portage-stable/profiles/package.mask
...
#>=dev-libs/boost-1.70.0
#>=dev-util/boost-build-1.70.0

```
In this way you can use _boost-1.75_ in the system with emerge.

```
emerge boost-build
emerge dev-libs/boost
```

Now it is possible to rebuild the system with all installed packages with emerge

`emerge -e world`

**Integrate boost-1.75.0 with the SDK**
After the HowTo  instructions for the _sdk_ integration with _boost-1.75_ you should edit  _/mnt/host/source/src/third_party/coreos-overlay/profiles/coreos/targets/sdk/package.accept_keywords_ and add the following line:
```
>=dev-utils/build-boost-1.75.0 ~amd64
>=dev-libs/boost-1.75.0 ~amd64
#>=dev-utils/build-boost-1.75.0 ~aarch64
#>=dev-libs/boost-1.75.0 ~aarch64
```
Then exit from chroot and as regular user run:
```
bin/cork enter
sudo FORCE_STAGES=stage4 ./bootstrap_sdk
...
>>> Emerging (1 of 2) dev-libs/boost-1.75.0::coreos
>>> Emerging binary (2 of 2) dev-util/boost-build-1.75.0::coreos
>>> Installing (2 of 2) dev-util/boost-build-1.75.0::coreos
>>> Installing (1 of 2) dev-libs/boost-1.75.0::coreos
...
INFO    bootstrap_sdk: Validating DIGESTS for flatcar-sdk-amd64-2605.9.0+2021-03-11-0633.tar.bz2.CONTENTS
flatcar-sdk-amd64-2605.9.0+2021-03-11-0633.tar.bz2.CONTENTS: OK
flatcar-sdk-amd64-2605.9.0+2021-03-11-0633.tar.bz2.CONTENTS: OK
flatcar-sdk-amd64-2605.9.0+2021-03-11-0633.tar.bz2.CONTENTS: OK
INFO    bootstrap_sdk: SDK ready: /mnt/host/source/src/build/catalyst/builds/flatcar-sdk/flatcar-sdk-amd64-2605.9.0+2021-03-11-0633.tar.bz2
INFO    bootstrap_sdk: Elapsed time (bootstrap_sdk): 206m2s
```
**References**
https://lists.boost.org/Archives/boost/2020/03/248336.php
https://forums.gentoo.org/viewtopic-t-1065178-start-0.html

**Comment by bonkko** [comment]
### Port of boost with minor modification from Gentoo(second draft)
From https://github.com/kinvolk/portage-stable/pull/153#issuecomment-796724935

**Differences with Gentoo**

**dev-util/boost-build-1.74.0**    _EAPI6_  and _variable name for the environment_ 
```
EAPI=6

inherit flag-o-matic prefix toolchain-funcs

MY_V="1.74.0"
MY_PV="1_74_0"

SRC_URI="https://dl.bintray.com/boostorg/release/${MY_V}/source/boost_${MY_PV}.tar.bz2"
```

**dev-util/boost-build-1.75.0**    _EAPI6_  and _variable name for the environment_ 

```
EAPI=6

inherit flag-o-matic prefix toolchain-funcs

MY_V="1.75.0"
MY_PV="1_75_0"

SRC_URI="https://dl.bintray.com/boostorg/release/${MY_V}/source/boost_${MY_PV}.tar.bz2"
```



**dev-libs/boost-1.74.0-r2** **    _EAPI6_ , _variable name for the environment_

```
EAPI=6

PYTHON_COMPAT=(python3_6)
IUSE="bzip2 context debug doc icu lzma -nls mpi numpy -python static-libs +threads tools zlib zstd"
MY_PV="1_74_0"
MAJOR_V="1_74_0"
```


**dev-libs/boost-1.75.0**    _EAPI6_ , _variable name for the environment_ , _-j1_ (with multiple job compilation fail in my machine)

```
EAPI=6

PYTHON_COMPAT=(python3_6)
IUSE="bzip2 context debug doc icu lzma -nls mpi numpy -python static-libs +threads tools zlib zstd"
MY_PV="1_75_0"
MAJOR_V="1_75_0"
src_configure() {
        OPTIONS=(
                "-j1"
                -q
                -d+2
                pch=off
                $(usex icu "-sICU_PATH=${ESYSROOT}/usr" '--disable-icu boost.locale.icu=off')
                $(usex mpi '' '--without-mpi')
                $(usex nls '' '--without-locale')
                $(usex context '' '--without-context --without-coroutine --without-fiber')
                $(usex threads '' '--without-thread')
                --without-stacktrace
                --prefix="${ED}/usr"
                --layout=system
                # CMake has issues working with multiple python impls,
                # disable cmake config generation for the time being
                # https://github.com/boostorg/python/issues/262#issuecomment-483069294
                # building with threading=single is currently not possible
                # https://svn.boost.org/trac/boost/ticket/7105
                threading=multi
                link=$(usex static-libs shared,static shared)
                # this seems to be the only way to disable compression algorithms
                # https://www.boost.org/doc/libs/1_70_0/libs/iostreams/doc/installation.html#boost-build
                -sNO_BZIP2=$(usex bzip2 0 1)
                -sNO_LZMA=$(usex lzma 0 1)
                -sNO_ZLIB=$(usex zlib 0 1)
                -sNO_ZSTD=$(usex zstd 0 1)
        )
        append-cxxflags -std=c++14
}
```
**Additional information**
In my opinion may be nice to:
- [ ] check compilation with multiple flag
- [ ] test on aarch64(native) and ppc32(native)   --- ppc32 is not supported by Flatcar but it should be easy to make a port
- [ ] make unit test for the code
- [ ]  check security impact on the code base
- [ ]  add support for Open MPI

I would like to stabilize and test other arch with different compile flags. 

**Current Status**
_Each step to make the suff listed below are described in the previous HowTo._

1. No error with
```
emerge boost-build
emerge dev-libs/boost
```
2. No error with
```
emerge -e world
```
 3.No error with SDK stage4 
```
bin/cork enter
sudo FORCE_STAGES=stage4 ./bootstrap_sdk
```
**Reference**
_https://gitweb.gentoo.org/repo/gentoo.git/_

**Comment by bonkko** [comment]
### Port of boost with minor modification from Gentoo (final draft)
This document describe the port of dev-libs/boost and dev-util/boost-build in portage-stable. EAPI 7 is used to follow original Gentoo ebuild(see https://github.com/kinvolk/portage-stable/pull/153/commits/7858867395caff1dd45d407831dc8c62b81b90a3).

**Differences with Gentoo**
_dev-libs/boost-1.74.0-r2_ **Python3_6** (see https://github.com/kinvolk/portage-stable/pull/153/commits/25ba533671542fbc36ccc4b154e279f8edceb706) 
_dev-libs/boost-1.75.0_ **Python3_6** (see https://github.com/kinvolk/portage-stable/pull/153/commits/25ba533671542fbc36ccc4b154e279f8edceb706) 

**Current Status**

- [X] No error with
```
emerge boost-build
emerge dev-libs/boost
```
- [X] No error with
```
emerge -e world
```
- [x] No error with SDK stage4 
```
bin/cork enter
sudo FORCE_STAGES=stage4 ./bootstrap_sdk
```
- [x] CI test passed (https://github.com/kinvolk/portage-stable/pull/153#issuecomment-801959247) 

**Additional information**
May be nice to:
- [ ] check compilation with multiple flag
- [ ] test on aarch64(native), ppc32(native) and ppc64(native)   --- ppc32 and ppc64 is not supported by Flatcar but it should be easy to make a port
- [ ]  make unit test for the code
- [ ]  check security impact on the code base
- [ ]  add support for Open MPI

**References**
https://github.com/kinvolk/portage-stable/pull/153

**Comment by margamanterola** [comment]
As mentioned on the PR, I've run CI against it and it build and run successfully on ARM. We currently don't support ppc, so I don't think that's necessary.

Regarding unit testing, it's not clear what you'd test, given that this is just building the newest boost version.  We run our tests using kola, they are more integration tests than unit tests. Our CI already did a run with the existing integration tests and they passed. Did you have something in mind of possible tests that we should add to have better boost coverage?

**Comment by bonkko** [comment]
Yes, in my opinion may be nice to test boost with the current ABI and make some security unit test in the system with packages that use boost. 

As second stuff we can add security unit test for the boost source code. It is an additional review to the boost library. The proposed flow is:  boost community write code and make a review ---> gentoo community make a review and check the code ---> we write additional unit test for the security of the library and we make additional code review and check.

**Comment by dongsupark** [comment]
https://github.com/kinvolk/portage-stable/pull/153 was merged, so boost 1.75.0 will be included in the next Alpha.
Thanks!

**Comment by bonkko** [comment]
I'm glad for it.
